scripts/kfctl.sh

```diff
@@ -8,6 +8,60 @@
 # cd myapp
 # kfctl.sh generate all
 # kfctl.sh apply all
+#
+# kfctl will assemble and deploy prototypes using ksonnet
+# Prototypes are grouped by component and are available
+# after installing their component
+#    `ks pkg install kubeflow/<component>`
+# Components can be grouped under modules by doing
+#    `ks module create <module-name>`
+#    `ks generate <prototype-name> <component-name> --module <module-name>`
+# One or more modules can be separately deployed by doing
+#    `ks env targets default --module <module-name> ...`
+#    `ks apply default`
+# ksonnet modules are used so that different combinations of prototypes 
+# can be grouped within a ksonnet environment. By setting the environment targets
+# as one or more modules the environment becomes composable either before deployment
+# or after. See `kfctl.sh add --help`.
+#
+# Syntax:
+# A module is named along with it's dependencies by doing
+#   `kfctl.sh add <module-name> --dependsOn <component-name> [<component-name>] ...`
```

Comment:
Why are we adding support for modules at kfctl.sh? If people want to manipulate modules why don't they do that directly using ks?

Reply:
Our current grouping is flat, we have one environment and one large set of components. If someone would like to add a new component without opening a PR in kubeflow they could do it by (for example)
```
ks registry add intel "intelai/openvino"
ks pkg install intel/openvino
ks generate openvino openvino
ks apply default -c openvino 
```
but this is very imperative and error-prone since the definition of what components are required in an openvino pipeline isn't well defined other than we can add an openvino component to an existing set of kubeflow components. Of the components that already exist we also don't know what prototypes were deployed. We can get this by doing a `ks env describe default` 
```
name: default
kubernetesversion: v1.10.7
path: default
destination:
  server: https://35.185.234.170
  namespace: kubeflow
targets:
- argo
- seldon
- katib
- core
- training
- serving
- notebooks
- pytorch
- openvino
libraries: {}
```
and even a `ks param list default`
```
COMPONENT         PARAM                               VALUE
=========         =====                               =====
argo              executorImage                       'argoproj/argoexec:v2.2.0'
argo              name                                'argo'
argo              uiImage                             'argoproj/argoui:v2.2.0'
argo              workflowControllerImage             'argoproj/workflow-controller:v2.2.0'
core.ambassador   ambassadorServiceType               'LoadBalancer'
core.ambassador   platform                            'null'
core.spartakus    reportUsage                         true
core.spartakus    usageId                             933944721
notebooks.jupyter platform                            'null'
ambassador        ambassadorImage                     'quay.io/datawire/ambassador:0.37.0'
ambassador        ambassadorServiceType               'ClusterIP'
...
```
but at this point it gets overwhelming...